kms-provider-kroxylicious-inmemory/src/main/java/io/kroxylicious/kms/provider/kroxylicious/inmemory/InMemoryKms.java

```diff
@@ -0,0 +1,230 @@
+import java.security.GeneralSecurityException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.KeyGenerator;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.GCMParameterSpec;
+
+import io.kroxylicious.kms.service.DekPair;
+import io.kroxylicious.kms.service.Kms;
+import io.kroxylicious.kms.service.KmsException;
+import io.kroxylicious.kms.service.Serde;
+import io.kroxylicious.kms.service.UnknownAliasException;
+import io.kroxylicious.kms.service.UnknownKeyException;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * An in-memory KMS to be used only for testing.
+ * Note that this exposes public methods that are not part of the {@link Kms} interface which are used for those
+ * KMS operations which are outside the scope of Kroxy itself (such as key provisioning).
+ */
+public class InMemoryKms implements
+        Kms<UUID, InMemoryEdek> {
+
+    private static final String WRAP_ALGO = "AES_256/GCM/NoPadding";
+    public static final String KEY_ALGO = "AES";
+    private final Map<UUID, SecretKey> keys;
+    private final KeyGenerator aes;
+    private final int numIvBytes;
+    private final int numAuthBits;
+    private final SecureRandom secureRandom;
+    private final Map<String, UUID> aliases;
+
+    private final AtomicInteger numDeksGenerated;
+
+    InMemoryKms(int numIvBytes, int numAuthBits,
+                Map<UUID, SecretKey> keys,
+                Map<String, UUID> aliases,
+                AtomicInteger numDeksGenerated) {
+        this.keys = new HashMap<>(keys);
+        this.aliases = new HashMap<>(aliases);
+        this.numDeksGenerated = numDeksGenerated;
+        this.secureRandom = new SecureRandom();
+        this.numIvBytes = numIvBytes;
+        this.numAuthBits = numAuthBits;
+        try {
+            this.aes = KeyGenerator.getInstance(KEY_ALGO);
+        }
+        catch (NoSuchAlgorithmException e) {
+            // This should be impossible, because JCA guarantees that AES is available
+            throw new KmsException(e);
+        }
+    }
+
+    /**
+     * Generates a KEK
+     * @return The id of the KEK
+     */
+    public UUID generateKey() {
+        var key = aes.generateKey();
+        var ref = UUID.randomUUID();
+        keys.put(ref, key);
+        return ref;
+    }
+
+    public void createAlias(UUID keyId, String alias) {
+        lookupKey(keyId); // check the key exists in this KMS
+        aliases.put(alias, keyId);
+    }
+
+    public void deleteAlias(String alias) {
+        var was = aliases.remove(alias);
+        if (was == null) {
+            throw new UnknownAliasException(alias);
```

Comment:
For how this is being used right now (i.e. in tests) I don't think it makes a huge difference either way. The value of it would be in detecting that your understanding about which aliases exist is incorrect. That actually seems useful to me in tests, where _not_ throwing could lead to a false positive (assertion passing which shouldn't). I'm inclined to leave this as is for now since it's only relevant to test code.